aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp

```diff
@@ -0,0 +1,225 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/core/op_registration/op_registration.h>
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <vector>
+
+namespace at {
+namespace native {
+namespace {
+
+inline int start_index(int a, int b, int c) {
+  return (int)std::floor((float)(a * c) / b);
+}
+
+inline int end_index(int a, int b, int c) {
+  return (int)std::ceil((float)((a + 1) * c) / b);
+}
+
+template <typename scalar_t, typename underlying_t>
+static void adaptive_avg_pool2d_single_out_frame(
+    scalar_t* input_p,
+    scalar_t* output_p,
+    int64_t sizeD,
+    int64_t isizeH,
+    int64_t isizeW,
+    int64_t osizeH,
+    int64_t osizeW,
+    int64_t istrideD,
+    int64_t istrideH,
+    int64_t istrideW) {
+  at::parallel_for(0, sizeD, 0, [&](int64_t start, int64_t end) {
```

Comment:
Loop order is poor for NHWC, you'll be jumping around reading one byte in multiples of C bytes

Reply:
Yes, I am implementing a fast path for NHWC layout. This is only for channel-first case.